core/src/main/scala/kafka/zookeeper/ZooKeeperClient.scala

```diff
@@ -75,6 +79,11 @@ class ZooKeeperClient(connectString: String,
   private[zookeeper] val reinitializeScheduler = new KafkaScheduler(threads = 1, s"zk-client-${threadPrefix}reinit-")
   private var isFirstConnectionEstablished = false
 
+  // There should never be more than one controller-failover call at a time, but no
+  // guarantees for end-user listTopics() calls, and zookeeper.max.in.flight.requests
+  // appears to be either 10 (default) or infinite, so go with 10 as safe bet:          
```

Comment:
"infinite"? zookeeper.max.in.flight.requests is an int, so between 0 and int.max. What is the possibility that there would be 10 listTopics at the same time? If it is low possibility, feel using a fixed of 10 threads is kind of wasting.

Reply:
Yeah, I'm not actually sure why I wrote that...the live configured value is 10 in my dev deployment, and 0 is disallowed by the code, so it doesn't count as "infinite."

Anyway, the probability is definitely low, but one could imagine a service doing a 100% bounce or deploy, with every instance coming up and trying to do one of these calls. What would you suggest as an alternative? While it's true there's some overhead to extra threads, if they're not doing anything, it's essentially zero CPU, and memory is mostly the stack, I think (which may default to 10 MB, but that's tiny relative to our deployed setup).

Reply:
I agree with 10 threads are not a significant pressure, especially when compared with the maintenance cost of debugging blocked requests on this, I'd say configured to the imaginable quota.

Reply:
Yes, I agree that 10 threads are not expensive. However, if we create unnecessary threads here and there, the accumulated usage could be huge. I am wondering whether controller has a shared threadpool that we could take threads from. If not, if the request rate is low, we could get a new thread each time for a new paginated request, so we do not need to have 10 threads waiting for most of the time. If the request rate is high and performance requirement is high, I guess having 10 threads should be fine.

Reply:
Controller has a thread pool dedicated for requests.  However this ZK part is a hidden component encapsulated underneath.  

If we'd like to control it better, maybe pull that to a upper level config instead of hard coding.